git-link.el

```diff
@@ -496,6 +513,12 @@ return (FILENAME . REVISION) otherwise nil."
       (git-link--read-remote)
     (git-link--remote)))
 
+(defun git-link--prepare-prefix (host format placeholder)
+  (if (or (string= "dev.azure.com" host)
+          (string-suffix-p ".visualstudio.com" host))
+      (concat "G" (if (eq 'branch format) "B" "C") placeholder)
+     placeholder))
```

Comment:
Hi, 

Endpoint specific handling should be encapsulated by the endpoint specific function. With the addition of `git-link--prepare-prefix`, the Azure specific handling is no longer encapsulated by `git-link-azure`. As a consequence, we have a new issue:

The above function does not consider `git-link-commit-remote-alist`. With this, one can direct `git-link-azure` be called  for any host, not just the defaults. So if one does:

```el
(add-to-list 'git-link-remote-alist '("somehost\\.com" git-link-azure)) 
```

The `git-link` function will still fail.

The new link generating conditionals already seems to have a lot of "gotchas": hence the checks using `memq`, `string-empty-p` and `null` (I do realize that this is in part do to excising in consistencies with return values). Now we're adding `throw` and `catch` and `dolist`. We've gone from 1 line to 7 or so.

I appreciate your work and I hate not being able to merge. The whole ideal of a preferred format list needs to be addressed holistically, from a higher level. It's clear that when we attempt to wedge it into the existing framework for link generation problems occur and code becomes more complex than it should be.

Reply:
My lack of testing here is also a problem. With proper tests, you'd get this feedback immediately instead of having to wait days for me :(

Reply:
It seems git-link need a refactor to solve your concerns, I'm unsure I'm on the right way, maybe your can draft a detailed design, or just implement it based on my changes.

Reply:
Yeah it's unfortunate but at least this list business came out of it which is a nice idea. Will do a refactor a some point for a 1.0 which is definitely overdue.

Reply:
Thanks for your hard work on this.